### ISSUE TITLE
Any chance support for mips64el can be upstreamed?

### ISSUE THREAD

**Issue by glaubitz** [loongson-community/grub#3]
While support for LoongArch was recently upstreamed in GRUB and will be available with version 2.12, the patches from this repository to support MIPS64EL were never upstreamed.

Is there a chance that these patches can be rebased and sent upstream?

**Comment by xry111** [comment]
It was submitted to upstream by @hev as https://lists.gnu.org/archive/html/grub-devel/2017-02/msg00043.html but completely ignored.

One difficulty is the implementation depends on the "Loongson Development System Firmware and Kernel Interface Specification" (sometimes called "BPI" for "Boot Protocol Interface") which is only used by Loongson MIPS-based processors.  It was intended to be submitted and standardized as "MIPS UEFI" but the standardization never happened and will not happen since Loongson has stepped away from MIPS to LoongArch for technical and commercial reasons.  So it would be very difficult to convince the upstream to accept such a bulk of vendor-specific code for a discontinued product line...

**Comment by xry111** [comment]
Note that one reason for removing IA64 support from GRUB is "it would be the only EFI platform not using EFI stub to boot the kernel".  So the upstream will strongly hate this unless we refactor it into something completely unrelated to EFI.

**Comment by glaubitz** [comment]
Thanks for the detailed explanation. I know the discussion around the ia64 removal as I was involved in it.

However, the GRUB upstream developers in charge have not expressed a very strong opinion on the issue if I remember correctly. It was more coming from the direction of some individual contributors.

We could ask Daniel Kiper for this opinion on the matter.

**Comment by xry111** [comment]
Let me try to rebase it first... GRUB < 2.12 won't boot kernels from ext4 partitions created using the defaults of e2fsprogs-1.47.0 because the support for metadata_csum_seed etc is only added in 2.12, so it's better to make GRUB 2.12 work anyway (upstreaming or not).

**Comment by xry111** [comment]
> Let me try to rebase it first... GRUB < 2.12 won't boot kernels from ext4 partitions created using the defaults of e2fsprogs-1.47.0 because the support for metadata_csum_seed etc is only added in 2.12, so it's better to make GRUB 2.12 work anyway (upstreaming or not).

I hit some troubles here.  Before rebasing I just tried the current code base.  I can successfully build it and boot into the GRUB shell.  But I cannot boot Linux from it.  I'm trying to debug but the debugging is difficult because after grub_efi_finish_boot_services we lose the serial output.

And my laptop is under repairing now so I'll just lay flat for several days :(.

**Comment by glaubitz** [comment]
OK, thanks for the update. Hope your laptop will be fixed soon.

Please let me know when there is anything I can help with.

**Comment by xry111** [comment]
https://github.com/xry111/grub/tree/xry111/2.12-lefi

But it does not work as at now.  Some issues:

1. The rebased code displays nothing on the screen.  I can only operate via serial console.
2. I still cannot boot the Linux kernel with it.
3. There are some code I don't know how to port correctly: https://github.com/xry111/grub/commit/950b9f656b664c37c3ce9a6e4f874c129cf9b5cf#diff-f63b4837337fdc1e8a94fb183c63282b625e4b41a54daa76c162489a778fcd4fR619

**Comment by glaubitz** [comment]
> https://github.com/xry111/grub/tree/xry111/2.12-lefi
> 
> But it does not work as at now. Some issues:
> 
>     1. The rebased code displays nothing on the screen.  I can only operate via serial console.
> 
>     2. I still cannot boot the Linux kernel with it.
> 
>     3. There are some code I don't know how to port correctly: [xry111@950b9f6#diff-f63b4837337fdc1e8a94fb183c63282b625e4b41a54daa76c162489a778fcd4fR619](https://github.com/xry111/grub/commit/950b9f656b664c37c3ce9a6e4f874c129cf9b5cf#diff-f63b4837337fdc1e8a94fb183c63282b625e4b41a54daa76c162489a778fcd4fR619)

Thank you very much for the update. I appreciate all the work.

**Comment by xry111** [comment]
I'm stepping back and trying to figure out the "cannot boot the kernel" issue which already happens in the 2.02 branch in this repo first.

I'll try to build on Loongnix (the "official" Loongson distro) and see if it'll work to rule out toolchain triggered issues.

**Comment by xry111** [comment]
If I replace `kernel.img` in my 2.02 custom build with the one from Loongnix, I can boot the kernel.  So whatever the bug is, it's in `kernel.img`.

**Comment by xry111** [comment]
I've tried building with the cross toolchain at http://ftp.loongnix.cn/toolchain/gcc/release/mips/gcc8/mips-loongson-gcc8-linux-gnu-2021-02-08-src.tar but no luck.  The Linux kernel still fails to build and the disassemble out is still so different beyond any attempt to compare.

**Comment by glaubitz** [comment]
> I've tried building with the cross toolchain at http://ftp.loongnix.cn/toolchain/gcc/release/mips/gcc8/mips-loongson-gcc8-linux-gnu-2021-02-08-src.tar but no luck. The Linux kernel still fails to build and the disassemble out is still so different beyond any attempt to compare.

Does the toolchain provided on kernel.org help? https://mirrors.edge.kernel.org/pub/tools/crosstool/

**Comment by xry111** [comment]
> > I've tried building with the cross toolchain at http://ftp.loongnix.cn/toolchain/gcc/release/mips/gcc8/mips-loongson-gcc8-linux-gnu-2021-02-08-src.tar but no luck. The Linux kernel still fails to build and the disassemble out is still so different beyond any attempt to compare.
> 
> Does the toolchain provided on kernel.org help? https://mirrors.edge.kernel.org/pub/tools/crosstool/

I guess not.  The disassembly of the "good" kernel.img contains some Loongson-specific instructions not even supported by the upstream toolchain, thus anything built with an upstream toolchain is likely still broken and absolutely not suitable for comparing the disassembly.

**Comment by xry111** [comment]
Next I'll try chroot'ing into the loongnix installation where the "good" kernel.img comes and building there.

**Comment by glaubitz** [comment]
Hmm, would it actually build with the loongarch64 toolchain?

**Comment by xry111** [comment]
> Hmm, would it actually build with the loongarch64 toolchain?

No, loongarch64 is a different ISA following the discontinue of Loongson MIPS processors.

And anyway my direction was completely wrong: it's not a toolchain issue.  I extracted 4 patches not in this repo from http://ftp.loongnix.cn/os/loongnix/1.0/SRPMS/g/grub2-2.02-0.40.fc21.loongson.12.src.rpm and now the kernel boots fine.  See https://github.com/xry111/grub/commits/xry111/2.02-lefi-1.

I'm not sure which changes in the 4 patches are really necessary though.

**Comment by glaubitz** [comment]
Wow, nice job! Thanks for getting this working!

**Comment by xry111** [comment]
A rebase to 2.12 also "works for me": ~~https://github.com/xry111/grub/releases/tag/grub-2.02-mips64-efi-2~~ https://github.com/xry111/grub/releases/tag/2.12-lefi-2

The screen display issue is also somehow fixed.

I've no idea if this will work for anyone else though.

**Comment by glaubitz** [comment]
Nice. I can give it a try on my Loongson mips64el machine as well.

**Comment by xry111** [comment]
> Nice. I can give it a try on my Loongson mips64el machine as well.

FYI initially I linked a wrong tag for the release.  Now corrected.

**Comment by xry111** [comment]
> > Nice. I can give it a try on my Loongson mips64el machine as well.
> 
> FYI initially I linked a wrong tag for the release. Now corrected.

And the URL should be https://github.com/xry111/grub/releases/tag/2.12-lefi-2

**Comment by Fearyncess** [comment]
the loongson-community's grub repo has no passing register $6 for linux kernel arguments.

https://github.com/chenx97/grub/tree/mips64-efi <- There is a fork that works well on most 3A4000/3B4000 machines. Based on https://github.com/yetist/grub

**Comment by xry111** [comment]
> the loongson-community's grub repo has no passing register $6 for linux kernel arguments.
> 
> https://github.com/chenx97/grub/tree/mips64-efi <- There is a fork that works well on most 3A4000/3B4000 machines. Based on https://github.com/yetist/grub

Please use it instead of my code.  Mingcong told me my code has a subtle bug causing initramfs not to work (I don't use an initramfs personally so I didn't notice the bug at all).

**Comment by glaubitz** [comment]
Maybe @chenx97 can upstream his patches so that mips64-efi will finally be supported natively by Debian.

**Comment by chenx97** [comment]
> Maybe @chenx97 can upstream his patches so that mips64-efi will finally be supported natively by Debian.

The firmware of my 3B4000 fails to map more than 256MB of memory. I wonder if I need to work around it before attempting to upstream my patches. Otherwise loading a regular kernel and an initramfs image may exceed this limit nowadays.

**Comment by Fearyncess** [comment]
<img width="1280" height="960" alt="Image" src="https://github.com/user-attachments/assets/0d81f048-7df8-46f7-b63b-fa248cb5e28a" />
another solution for non-LEFI systems

**Comment by xry111** [comment]
Applied @chenx97's patches for 2.14 at https://github.com/xry111/grub/releases/tag/2.14-lefi-1.

**Comment by glaubitz** [comment]
> Applied [@chenx97](https://github.com/chenx97)'s patches for 2.14 at https://github.com/xry111/grub/releases/tag/2.14-lefi-1.

Any chance someone could try to upstream this series? At least posting the series for a review on the grub-devel mailing list would be great.
